File: 0x-bonus/yegge/largest_int/largest_int.rb
a = Array.new(1000) { rand(1..1000) }
# Find the largest int value in an int array.

# Straightfoward with iteration
def largest(integers)
  max = -1.0 / 0.0 # Negative Infinity

  integers.each do |n|
    max = n if n > max
  end

  max
end

puts a.max
puts largest(a)

# Reduce / Inject
# Recursion

File: 0x-bonus/yegge/largest_int/largestInt.js
const a = [3, 76, 83, 20, 97, 23, 19, 34, 87, 30, 92, 100, 94, 78, 13, 86, 51, 83, 47, 90]

// TODO: recursive solution
// const largest = (integers) => {
//   let max = -Infinity;
//
//   integers.forEach((n) => {
//     if (n > max) {
//       max = n;
//     }
//   });
//
//   return max;
// };

// You need a lot of coffee to understand this.
// const largest = (integers) => integers.reduce((max, n) => n > max ? n : max);

const largest = (integers) => {
  return integers.reduce((max, n) => {
    if (n > max) {
      return n;
    } else {
      return max;
    }
  });
}


console.log( Math.max( ...a ) ); // Math.max expects a list not an array so we need to use the spread operator.
console.log( largest(a) );
